terraform-modules/teleport-aws-discovery/main.tf

```diff
@@ -0,0 +1,287 @@
+locals {
+  create                = var.create
+  discover_organization = false # TODO(gavin): impl org discovery
+  name_prefix = (
+    var.name_prefix != ""
+    ? "${trimsuffix(var.name_prefix, "-")}-"
+    : ""
+  )
+  tags = merge(var.tags, {
+    "teleport.dev/cluster"     = local.teleport_cluster_name
+    "teleport.dev/integration" = local.teleport_integration_name
+    # this is the origin we set for resources created by the AWS OIDC integration web UI wizard.
+    "teleport.dev/origin" = "integration_awsoidc"
+  })
+  teleport_cluster_name      = var.teleport_cluster_name
+  teleport_proxy_public_addr = var.teleport_proxy_public_addr
+  teleport_proxy_public_url  = "https://${local.teleport_proxy_public_addr}"
+}
+
+################################################################################
+# AWS IAM OIDC Provider
+################################################################################
+
+locals {
+  create_aws_oidc_provider  = local.create
+  aws_iam_oidc_provider_aud = "discover.teleport"
+  # strip the port since AWS OIDC provider doesn't support port in the url
+  aws_iam_oidc_provider_url = replace(local.teleport_proxy_public_url, "/:[0-9]+.*/", "")
+  default_aws_resource_name = "teleport-discovery"
+}
+
+data "tls_certificate" "teleport_proxy" {
+  count = local.create_aws_oidc_provider ? 1 : 0
+
+  url = local.teleport_proxy_public_url
+}
+
+# Create an AWS OIDC Provider, so that the Teleport Discovery Service can use
+# OIDC to assume the discovery AWS IAM role.
+resource "aws_iam_openid_connect_provider" "teleport" {
+  count = local.create_aws_oidc_provider ? 1 : 0
+
+  url             = local.aws_iam_oidc_provider_url
+  client_id_list  = [local.aws_iam_oidc_provider_aud]
+  thumbprint_list = [data.tls_certificate.teleport_proxy[0].certificates[0].sha1_fingerprint]
+  tags            = local.tags
+}
+
+################################################################################
+# AWS IAM role for Teleport Discovery Service
+################################################################################
```

Comment:
What do you think about splitting the module into separate .tf files to logically group resources it creates? Right now it's all in the same main.tf but I think it would be easier to understand the module if it was organized into files like `aws_roles.tf`, `aws_oidc_provider.tf`, `teleport_token.tf`, `teleport_discovery_config.tf` etc. I think that's a pretty common way to organize tf code.